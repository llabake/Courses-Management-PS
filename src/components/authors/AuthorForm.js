import React from 'react';
import {
  string,
  shape,
  arrayOf,
  bool,
  func,
  object,
} from 'prop-types';
import TextInput from '../common/TextInput';

const AuthorForm = ({
  author,
  onSave,
  onChange,
  saving,
  errors,
}) => (
  <form>
    <h1>Manage Authors</h1>
    <TextInput
        name="firstName"
        label="First Name"
        value={author.firstName}
        onChange={onChange}
        error={errors.firstName}
      />

    <TextInput
      name="lastName"
      label="Last Name"
      value={author.lastName}
      onChange={onChange}
      error={errors.lastName}
      />

    <input
        type="submit"
        disabled={saving}
        value={saving ? 'Saving' : 'Save'}
        className="btn btn-primary"
        onClick={onSave}
      />
  </form>
);

export default AuthorForm;
