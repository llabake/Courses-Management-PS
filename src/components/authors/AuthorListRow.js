/* eslint-disable jsx-a11y/no-noninteractive-element-interactions */
/* eslint-disable jsx-a11y/click-events-have-key-events */
import React from 'react';
import PropTypes from 'prop-types';
import { Link } from 'react-router-dom';

const AuthorListRow = ({ author, onDelete }) => {
  const deleteAuthor = () => (
    onDelete(author.id)
  );
  return (
    <tr>
      <td><Link to={`/author/${author.id}`}>{author.id}</Link></td>
      <td>
        {author.firstName} 
        {' '}
        {author.lastName}
      </td>
      <td
        className="btn btn-xs btn-danger"
        onClick={deleteAuthor}
      >
        Delete
      </td>
    </tr>
  );
};

AuthorListRow.propTypes = {
  author: PropTypes.object.isRequired,
  onDelete: PropTypes.func.isRequired,
};

export default AuthorListRow;
