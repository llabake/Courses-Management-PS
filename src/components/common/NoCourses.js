import React from 'react';
import PropTypes from 'prop-types';

const NoCourses = ({ text }) => (
  <div className="text-center">
    <h4>
      {text}
    </h4>
  </div>
);

NoCourses.propTypes = {
  text: PropTypes.string.isRequired,
};

export default NoCourses;
