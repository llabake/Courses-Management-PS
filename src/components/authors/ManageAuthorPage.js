import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Prompt } from 'react-router-dom';
import toastr from "toastr";
import {
  bool,
  shape,
  arrayOf,
  func,
} from 'prop-types';
import { authorInputValidator } from "../../helpers";
import { loadAuthors, saveAuthor } from "../../actions/authorActions";
import AuthorForm from "./AuthorForm";

export class ManageAuthorPage extends Component {
  state = {
    author: Object.assign({}, this.props.author),
    errors: {},
    saving: false,
    isBlocking: false,
    isValid: false,
  };

  updateAuthorState = (event) => {
    const field = event.target.name;
    const { author } = this.state;
    author[field] = event.target.value;
    return this.setState({ author, isBlocking: true });
  };

  saveAuthor = (event) => {
    event.preventDefault();
    if (!this.validate()) {
      return;
    }
    this.setState({ saving: true });
    this.props.saveAuthor(this.state.author)
      .then(() => this.redirect())
      .catch((error) => {
        toastr.error(error);
        this.setState({ saving: false });
      });
  };

  validate() {
    const { errors, isValid } = authorInputValidator(this.state);
    this.setState({ isValid, errors });
    return isValid;
  }

  redirect() {
    this.setState({
      saving: false,
      isBlocking: false,
    });
    toastr.success("Author saved successfully");
    this.props.history.push('/authors');
  }

  render() {
    const {
      author, errors, isBlocking, saving, 
    } = this.state;
    return (
      <div>
        <Prompt
          when={isBlocking}
          message="Are you sure you want to go to leave the page"
        />
        <AuthorForm
          author={author}
          errors={errors}
          onSave={this.saveAuthor}
          saving={saving}
          onChange={this.updateAuthorState}
        />
      </div>
    );
  }
}

export const mapStateToProps = (state, { match: { params } }) => {
  const authorId = params.id;
  let author = {
    id: '', firstName: '', lastName: '',
  };

  if (authorId && state.authors.length > 0) {
    author = state.authors.find(authorObj => authorObj.id === authorId);
  }

  return {
    author,
    authors: state.authors,
  };
};

export const mapDispatchToProps = dispatch => ({
  saveAuthor: author => dispatch(saveAuthor(author)),
});

export default connect(mapStateToProps, mapDispatchToProps)(ManageAuthorPage);
