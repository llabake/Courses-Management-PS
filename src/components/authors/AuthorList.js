import React from 'react';
import PropTypes from 'prop-types';
import AuthorListRow from "./AuthorListRow";

const AuthorList = ({ authors, onDelete }) => (
  <table className="table">
    <thead>
      <tr>
        <th>ID</th>
        <th>Name</th>
        <th />
      </tr>
    </thead>
    <tbody>
      {authors.map(author => (
        <AuthorListRow
          key={author.id}
          author={author}
          onDelete={onDelete} />
      ))}
    </tbody>
  </table>
);

AuthorList.propTypes = {
  authors: PropTypes.array.isRequired,
  onDelete: PropTypes.func.isRequired,
};

export default AuthorList;
