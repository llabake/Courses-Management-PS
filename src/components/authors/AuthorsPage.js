import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import toastr from "toastr";
import NoCourses from "../common/NoCourses";
import AuthorList from "./AuthorList";
import { loadAuthors, deleteAuthor } from "../../actions/authorActions";

class AuthorsPage extends Component {
  static authorRow(author, index) {
    return <div key={index}>{author.id}</div>;
  }

  state = {
    author: {},
  };

  redirectToAddAuthorPage = () => {
    const { history } = this.props;
    history.push('/author');
  };

  authorCourseExist = (courses, authorId) => {
    return courses.some(course => (course.authorId === authorId));
  };

  deleteAuthor = (authorId) => {
    const { history, courses } = this.props;
    if (!this.authorCourseExist(courses, authorId)) {
      this.props.deleteAuthor(authorId).then(() => {
        toastr.success('Author deleted successfully');
        history.push('/authors');
      });
    } else {
      toastr.warning(`This author can not be deleted as they have a course`);
    }
  };

  render() {
    const { authors, loading } = this.props;
    return (
      <div>
        <h1>Authors</h1>
        <input
          type="submit"
          value="Add Author"
          className="btn btn-primary"
          onClick={this.redirectToAddAuthorPage}
        />
        { !loading && !authors.length
          ? (
            <NoCourses
              text="There are no Authors now. Click the Add Author button to add an author" />
          )
          : (
            <AuthorList
            onDelete={this.deleteAuthor}
            authors={authors}
          />
          )
        }
      </div>
    );
  }
}

export const mapStateToProps = state => ({
  courses: state.courses,
  loading: state.ajaxCallsInProgress > 0,
  authors: state.authors,
});

export const mapDispatchToProps = dispatch => ({
  deleteAuthor: author => dispatch(deleteAuthor(author)),
});

export default connect(mapStateToProps, mapDispatchToProps)(AuthorsPage);
