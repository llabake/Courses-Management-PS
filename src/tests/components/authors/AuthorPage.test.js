import React from 'react';
import { shallow, mount } from 'enzyme';
import NoCourses from "../../../components/common/NoCourses";
import { AuthorsPage, mapDispatchToProps } from "../../../components/authors/AuthorsPage";
import AuthorList from "../../../components/authors/AuthorList";
import authorMock from "../../mocks/authorMock";
import courseMock from "../../mocks/courseMock";

describe('<AuthorsPage', () => {
  it('renders correctly', () => {
    const state = {
      author: {},
    };
    const props = {
      courses: [],
      loading: false,
      deleteAuthor: jest.fn(),
      authors: [],
    };
    const wrapper = shallow(<AuthorsPage {...props} {...state} />);
    expect(wrapper).toMatchSnapshot();
  });
  it('renders No courses component when there are no authors', () => {
    const state = {
      author: {},
    };
    const props = {
      courses: [],
      loading: false,
      deleteAuthor: jest.fn(),
      authors: [],
    };
    const wrapper = shallow(<AuthorsPage {...props} {...state} />);
    expect(wrapper.find(NoCourses)).toBeTruthy();
  });
  it('renders a list of courses when there are courses', () => {
    const props = {
      authors: authorMock.authors,
      loading: false,
      deleteAuthor: jest.fn(),
      redirectToAddAuthorPage: jest.fn(),
      courses: courseMock.courses,
    };
    const wrapper = shallow(<AuthorsPage {...props} />);
    expect(wrapper.find(AuthorList)).toBeTruthy();
  });
  it('redirects to add course page', () => {
    const props = {
      courses: courseMock.courses,
      authors: authorMock.authors,
      loading: false,
      deleteAuthor: jest.fn(),
      redirectToAddAuthorPage: jest.fn(),
      history: { push: jest.fn() },
    };

    const wrapper = shallow(<AuthorsPage {...props} />);
    wrapper.find('.btn').simulate('click');
    expect(props.history.push).toHaveBeenCalledWith('/author');
  });
  it('delete a author ', () => {
    const props = {
      courses: courseMock.courses,
      authors: authorMock.authors,
      loading: false,
      deleteAuthor: jest.fn(() => Promise.resolve()),
      redirectToAddAuthorPage: jest.fn(),
      history: { push: jest.fn() },
    };

    const wrapper = shallow(<AuthorsPage {...props} />);
    const spy = jest.spyOn(wrapper.instance(), 'deleteAuthor');
    wrapper.instance().deleteAuthor('mock-author-one');
    expect(spy).toHaveBeenCalled();
  });

  it('should dispatch needed actions', () => {
    const dispatch = jest.fn();
    expect(mapDispatchToProps(dispatch)).toHaveProperty('deleteAuthor');
    const { deleteAuthor } = mapDispatchToProps(dispatch);
    deleteAuthor();
    expect(dispatch).toHaveBeenCalled();
  });
});
