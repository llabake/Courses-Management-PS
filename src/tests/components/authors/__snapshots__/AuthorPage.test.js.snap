// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AuthorsPage renders correctly 1`] = `
<div>
  <h1>
    Authors
  </h1>
  <input
    className="btn btn-primary"
    onClick={[Function]}
    type="submit"
    value="Add Author"
  />
  <NoCourses
    text="There are no Authors now. Click the Add Author button to add an author"
  />
</div>
`;
